de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/concurrent/management/JupiterClient.java

```diff
@@ -71,4 +78,30 @@ public synchronized ChecksumActivity withTimestamp(ChecksumActivity checksumActi
 
     return get(checksumActivity.getPath()).withTimestamp(checksumActivity);
   }
+
+  @Override
+  public void start() {
+    jupiterHeartbeat =
```

Comment:
I am not sure how `ScheduledThreadPoolExecutor` is used correctly.
Is it necessary to create a new instance every time start is called?
If the same object could be re-used, would there a reason not to instantiate the executor in the constructor and just to re-schedule the periodic execution here?

Reply:
From the docs of the `shutdown` method (called in `stop` to "un"-schedule the heartbeat):
> Initiates an orderly shutdown in which previously submitted tasks are executed, but no new tasks will be accepted.

The `getContinueExistingPeriodicTasksAfterShutdownPolicy`-method does suggest, that this also stops repeating tasks:
> Gets the policy on whether to continue executing existing periodic tasks even when this executor has been shutdown. In this case, these tasks will only terminate upon shutdownNow or after setting the policy to false when already shutdown. This value is by default **false**.

`scheduleWithFixedDelay` does return a `ScheduledFuture`, which can be cancelled, but I am not sure, how that works in the context of a fixed rate (the other `schedule`-methods also return such futures). Can this future resolve multiple times or does it never fully resolve? Does it only cancel the current/next tick or does it cancel all of them?

The documentation is also not particular clear on these questions.
E.g. is cancelling the future the same as cancelling the executor? (earlier they only hinted at **shutting down the executor** is this the same?) (also this is the only reference to cancel in the method description):
> If any execution of the task encounters an exception, subsequent executions are suppressed. Otherwise, the task will only terminate via **cancellation** or termination of the **executor**.

Or how nothing is said about cancellation regarding the return value (the future):
> a ScheduledFuture representing pending completion of the task, and whose get() method will throw an exception upon cancellation.

To me using `shutdown` seemed easier then to test the specific nature of this beast.

Reply:
@stefaus mentioned, that he probably has some knowledge regarding this discussion point. I would love to make this kinda scary patch simpler.

Reply:
I think the solution now is good.
At some time you need to stop the Thread Pool or the JVM will wait forever.
A cancel on the Future or remove wouldn't stop the Thread Pool.

And if you get a Exception in `dispatchHeartbeats` (which will now **silently** cancel all following schedules..., yes no stacktrace) this will be running in the next Session again.
Mentioning this, you should wrap the code of `dispatchHeartbeats` in a try-catch and write exceptions to the log.